### ISSUE TITLE
Unable to publish to collaboration target via RStudio IDE

### ISSUE THREAD

**Issue by toph-allen** [rstudio/rstudio#15065]
When publishing to Connect from the RStudio IDE, if the deployment target is a collaboration, the content is published to the currently-registered user's account instead. The replication steps for this one are pretty long.

This _may_ be a bug in the RStudio IDE, but a little more investigation is needed first.

### Phase 1.

Start with an IDE with no accounts registered and an unpublished Shiny app.

Register the first account, here named `publisher`. Publish the content item.

![Screen Shot 2024-07-19 at 04 51 17 PM@2x](https://github.com/user-attachments/assets/3df1290d-b84d-4459-83ca-ffb809b58ee3)

You now have a DCF file under `content/rsconnect/server/publisher/shinyapp.dcf`

![Screen Shot 2024-07-19 at 04 52 36 PM@2x](https://github.com/user-attachments/assets/9b6a8f0a-4fa9-44e2-b9f7-a3cda82dcb3a)

Add another user as a collaborator. In this case, the user is named `toph`.

Disconnect the original user from RStudio, quit, and re-open the application.

Register the second user's account in RStudio.

RStudio detects one destination, from the `publisher/shinyapp.dcf` file. (We can prove this because when the dir and file are removed, the destination disappears. They reappear when the files are replaced)

Publish the content to this destination.

![Screen Shot 2024-07-19 at 04 57 17 PM@2x](https://github.com/user-attachments/assets/c1a7c961-4bc1-4c29-86d5-3d2795ab7659)

![Screen Shot 2024-07-19 at 04 57 33 PM@2x](https://github.com/user-attachments/assets/88bf1a49-853d-481e-b2ee-61bfdc5b095d)

Expected:

- The application is updated

Actual:

- A new application is created under the currently-registered user.
- Trying to publish again to the `publisher` destination warns about overwriting the `toph` destination.

![Screen Shot 2024-07-19 at 04 58 13 PM@2x](https://github.com/user-attachments/assets/55a93560-98e3-4268-90f3-32937585c642)
![Screen Shot 2024-07-19 at 04 59 43 PM@2x](https://github.com/user-attachments/assets/cfdbb388-ec82-425b-b57e-c10d0e6a74fa)

### Phase 2.

Delete `toph/shinyapp.dcf` which represented the second destination. 

Run `rsconnect::deployApp(appId = GUID)`, where `GUID` is the GUID of the intended collaboration target, to force deployment to the correct location.

The expected thing happens:

- The deployment goes to the right place.
- A new DCF file is created, this time pointing to the `publisher` destination, as seen by `username: publisher`.

![Screen Shot 2024-07-19 at 05 02 55 PM@2x](https://github.com/user-attachments/assets/04effce5-34df-490e-85a6-ba9a23fe235b)

Now, in the publish dropdown menu, there are two destinations.

![Screen Shot 2024-07-19 at 05 03 09 PM@2x](https://github.com/user-attachments/assets/af450f0b-0892-4e12-9a2d-3c619d1ed5b9)

Each of them is generated from a different DCF file, and show slightly different screens in the deployment wizard. Screenshots of those are not included, because *neither* of them deploy to the correct destination -- they both deploy to the same appname under the user account of the currently-registered user, instead of the collaboration destination.

I can't tell yet if this is an `rsconnect` or an IDE bug. It feels like there is a bug with the display of DCF files, and the extraction of targets from DCF files.

My understanding was that the expected behavior was: Only show DCF files for the current user. Allow a collaboration target to be added by using `deployApp(appID = GUID)`.

Actual behavior of IDE menu: Show DCF files from all accounts. No deployment targets allow the publication to content items under someone else's account.

Recent changes in RStudio which may be related to this:
- https://github.com/rstudio/rstudio/pull/14178
- https://github.com/rstudio/rstudio/pull/14486

**Comment by lachlansimpson** [comment]
Seen in  ticket #108789

**Comment by lachlansimpson** [comment]
Seen in ticket #109584

**Comment by leesahanders** [comment]
seen in ticket #110478

**Comment by gtritchie** [comment]
Marking as high priority since multiple tickets mention this issue.

**Comment by gtritchie** [comment]
Confirmed that the regression was introduced by https://github.com/rstudio/rstudio/pull/14178.

**Comment by kevinushey** [comment]
Thanks for doing the legwork @gtritchie. I can confirm the issue here comes down to `rsconnect::applications()` not including these applications in the returned list, so the move from `rsconnect:::getAppById()` to `rsconnect::applications()` is problematic.

We made this change because we wanted to present the `config_url` in the UI for the publish dialog, but that URL isn't returned via `rsconnect:::getAppById()`.

@toph-allen, would it be possible for Connect to return `config_url` for these applications in the `getApplication()` client request?

**Comment by kevinushey** [comment]
Hmm, apparently that's just constructed "by hand" from the application / server URL here:

https://github.com/rstudio/rsconnect/blob/2fe8fdef45f843cfb1316fecb443ac30a0ed8a42/R/applications.R#L98-L108

so we could just do the same thing ourselves (although needing to duplicate code is a bit unfortunate)

**Comment by gtritchie** [comment]
Verified using build 2024.11.0+345. Created a shiny app in a shared folder, then deployed it as user "fred". Added user "tom" as a collaborator.

"tom" then opened the same shiny app (from the shared folder, so it has the rsconnect folder created when "fred" published), made a small tweak, then republished.

Confirmed that the existing deployment was updated, instead of a new one being created.

I also reverified https://github.com/rstudio/rstudio/issues/12708, since the fix for that is what originally triggered this regression. Confirmed it is still working as intended.
